api/envoy/api/v2/cds.proto

```diff
@@ -234,6 +234,32 @@ message Cluster {
   //   verification.
   auth.UpstreamTlsContext tls_context = 11;
 
+  // UpstreamTransportSocket specifies the transport socket config used for upstream connection.
+  message UpstreamTransportSocket {
+    oneof socket_config {
+      core.TransportSocket transport_socket = 1;
+      auth.UpstreamTlsContext tls_context = 2;
+    }
+  }
+
+  // Configuration to use different transport sockets for different endpoints.
+  // This should only be used for EDS typed cluster.
+  // The value of LbEndpoint.Metadata configuration under the *envoy.transport_socket_matcher*
+  // is used to decide which transport socket to use when connecting to that particular endpoint.
+  // For example, an endpoint with metadata
+  //
+  // .. code-block:: json
+  //
+  //   {
+  //     "envoy.transport_socket_matcher_label":
+  //        { "label": "mtlsEnabled" }
+  //   }
+  //
+  // This endpoint will use the transport_socket_matcher["mtlsEnabled"] configuration.
+  // In case of of label is missing or the transport_socket_matcher does not have corresponding
+  // entry in the map, Envoy fallbacks to the *tls_context* or *transport_socket* in the cluster.
+  map<string, UpstreamTransportSocket> transport_socket_matcher = 41;
```

Comment:
I'm hesitating of terminology, `selector` might be a better term?

Reply:
This seems inconsistent from the proposal doc in #8016, why this is a map? cc @PiotrSikora

Reply:
okay, just realized the doc is using the repeated list.

In that case, we should define `UpstreamTransportSocket` to include a field like `string match`.

And then replace this map with

`repeated UpstreamTransportSocket transport_socket_matcher = 41`.

Does that sound better? I can't tell how it's better than map or any issue with the map though. I mean the internal implementation we still need to have a map for lookup.

Reply:
Take another look

We already have the [`TransportSocket`](https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/core/base.proto#L218)
defined.

We already have implementation to the `tls_context` to this struct in [`upstream_impl`](https://sourcegraph.com/github.com/envoyproxy/envoy@master/-/blob/source/common/upstream/upstream_impl.cc?suggestion#L727)

Maybe we can just use define it as

```proto
message Cluster {
  message UpstreamTransportSocketMatch {
     string endpoint_label_match;
     core.TransportSocket transport_socket;
  }
  repeated UpstreamTransportSocketMatch transport_socket_matcher;
}
```

Then the xDS server thrust `TlsContext` config into the `core.TransportSocket.Config` typed as struct.Any.

Basically I guess same as what proposed in #8098 get rid of the oneof.

Reply:
yes, UpstreamTlsContext in transport_socket.config or typed_config should already work

Reply:
Resolving.